nifi-nar-bundles/nifi-websocket-bundle/nifi-websocket-processors/src/main/java/org/apache/nifi/processors/websocket/AbstractWebSocketGatewayProcessor.java

```diff
@@ -0,0 +1,314 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.websocket;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.nifi.annotation.behavior.TriggerSerially;
+import org.apache.nifi.annotation.lifecycle.OnStopped;
+import org.apache.nifi.annotation.lifecycle.OnUnscheduled;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.components.ValidationContext;
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.flowfile.attributes.CoreAttributes;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.websocket.BinaryMessageConsumer;
+import org.apache.nifi.websocket.ConnectedListener;
+import org.apache.nifi.websocket.TextMessageConsumer;
+import org.apache.nifi.websocket.WebSocketClientService;
+import org.apache.nifi.websocket.WebSocketConfigurationException;
+import org.apache.nifi.websocket.WebSocketConnectedMessage;
+import org.apache.nifi.websocket.WebSocketMessage;
+import org.apache.nifi.websocket.WebSocketService;
+import org.apache.nifi.websocket.WebSocketSessionInfo;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.LinkedBlockingQueue;
+
+@TriggerSerially
+public abstract class AbstractWebSocketGatewayProcessor extends AbstractWebSocketProcessor implements ConnectedListener, TextMessageConsumer, BinaryMessageConsumer {
+
+    public static final PropertyDescriptor PROP_MAX_QUEUE_SIZE = new PropertyDescriptor.Builder()
+            .name("max-queue-size")
+            .displayName("Max Queue Size")
+            .description("The WebSocket messages are kept in an on-memory queue," +
+                    " then transferred to relationships when this processor is triggered." +
+                    " If the 'Run Schedule' is significantly behind the rate" +
+                    " at which the messages are arriving to this processor then a back up can occur." +
+                    " This property specifies the maximum number of messages this processor will hold in memory at one time." +
+                    " CAUTION: Any incoming WebSocket message arrived while the queue being full" +
+                    " will be discarded and a warning message will be logged.")
+            .required(true)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("10000")
+            .build();
+
+    private volatile LinkedBlockingQueue<WebSocketMessage> incomingMessageQueue;
```

Comment:
This particular issue keeps coming up. Basically the pattern of maintaining an internal queue for messages may lead to OOM exception as well as lost data if processor crashes while the queue still has data. This is due to the fact that you are essentially integrating with external scheduling mechanism
One way of mitigating this is to rely on such mechanism for distributing messages while essentially ignoring our schedule. You can look at the example of it in SMTP processor https://github.com/apache/nifi/blob/d838f61291d2582592754a37314911b701c6891b/nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ListenSMTP.java#L175
While may not be ideal, it certainly paves a simpler path for future improvement when we get a chance to define an Event-driver Processor abstraction as a first class component of NiFi.